### ISSUE TITLE
Information regarding WSL2 changes

### ISSUE THREAD

**Issue by billziss-gh** [microsoft/WSL2-Linux-Kernel#1]
Please consider adding information regarding Microsoft's WSL2 changes to Linux in your README.

**Comment by sarathyweb** [comment]
2001: Microsoft CEO:
" Linux is a cancer"
https://www.theregister.co.uk/2001/06/02/ballmer_linux_is_a_cancer/
2019: "Windows ships with a Linux kernel"
https://github.com/microsoft/WSLv2-Linux-Kernel

**Comment by edjroot** [comment]
So... The cancer has spread!

**Comment by sarathyweb** [comment]
Definitely

**Comment by glixlur** [comment]
It’s almost like companies can change, right?

**Comment by nextgenthemes** [comment]
> 2001: Microsoft CEO:
> " Linux is a cancer"
> https://www.theregister.co.uk/2001/06/02/ballmer_linux_is_a_cancer/
> 2019: "Windows ships with a Linux kernel"
> https://github.com/microsoft/WSLv2-Linux-Kernel

Why do you rons downvote this? Its nothing but a true statement.

**Comment by tills13** [comment]
> Why do you rons downvote this? Its nothing but a true statement.

This is not reddit. Let's keep commentary here relevant.

**Comment by sashalevin** [comment]
Hi Bill,

We have a very small amount of changes to the kernel, and most of them are commits that have went upstream but not backported to the 4.19 LTS tree. Out plan is to stay as close as possible to the latest kernel LTS tree.

Is there any change in particular you were looking for more information on?

**Comment by billziss-gh** [comment]
@sashalevin thanks for your answer.

I am assuming that Microsoft had to make some changes to the Linux kernel to optimize it for use in the WSL2 scenarios. This is based on prior public information from Microsoft.

For example: [[link](https://devblogs.microsoft.com/commandline/shipping-a-linux-kernel-with-windows/)]

> In addition to the LTS source from Kernel.org, a number of local patches are being applied. These patches tune the resulting binary for use in WSL2 by improving launch times, reducing the memory footprint and curating a minimal set of supported devices. The result is a small, lightweight kernel that is purpose built for WSL2 to be a drop-in replacement for the emulation architecture featured in the design of WSL1.

It is not immediately obvious (to me) what these changes are. If there are no major changes, or if the changes are minor configuration changes, perhaps a brief writeup describing some of those changes might suffice.

**Comment by sashalevin** [comment]
In general, these optimizations are making their way upstream, there's no blob of "secret sauce" in here. There are mostly commits that allow us to build a smaller kernel (such as https://github.com/microsoft/WSL2-Linux-Kernel/commit/22e33e1f5cc204930fb13eea012d52b5be13b77f) or commits that improve hyper-v and can't be backported to the LTS tree because they're not fixes (such as https://github.com/microsoft/WSL2-Linux-Kernel/commit/978f68a42df27a278e69460a3f4b006757ed8ffd).

The list of those commits is available by diffing the upstream LTS tree with the master branch of this repo.

**Comment by aidanharris** [comment]
Just for fun, it looks like this is based off of the 4.19.52 kernel ([Makefile](https://github.com/microsoft/WSL2-Linux-Kernel/blob/7fec2bd4a7fd7a952e3e5ea2202bef963aa4781d/Makefile#L4)):

```bash
mkdir linux
cd linux
git clone --depth 1 --branch v4.19.52 https://github.com/gregkh/linux.git upstream
git clone --depth 1 https://github.com/microsoft/WSL2-Linux-Kernel wsl2
diff -qr --exclude=.git upstream wsl2
```

```
Files upstream/arch/x86/include/asm/pci_x86.h and wsl2/arch/x86/include/asm/pci_x86.h differ
Files upstream/drivers/acpi/acpica/dsutils.c and wsl2/drivers/acpi/acpica/dsutils.c differ
Files upstream/drivers/acpi/internal.h and wsl2/drivers/acpi/internal.h differ
Files upstream/drivers/acpi/Kconfig and wsl2/drivers/acpi/Kconfig differ
Files upstream/drivers/acpi/Makefile and wsl2/drivers/acpi/Makefile differ
Files upstream/drivers/acpi/nfit/core.c and wsl2/drivers/acpi/nfit/core.c differ
Files upstream/drivers/hv/channel_mgmt.c and wsl2/drivers/hv/channel_mgmt.c differ
Files upstream/drivers/pci/Makefile and wsl2/drivers/pci/Makefile differ
Files upstream/drivers/tty/sysrq.c and wsl2/drivers/tty/sysrq.c differ
Files upstream/include/acpi/acpi_drivers.h and wsl2/include/acpi/acpi_drivers.h differ
Files upstream/include/linux/acpi.h and wsl2/include/linux/acpi.h differ
Files upstream/include/linux/pci.h and wsl2/include/linux/pci.h differ
Files upstream/include/uapi/linux/ndctl.h and wsl2/include/uapi/linux/ndctl.h differ
Files upstream/init/Kconfig and wsl2/init/Kconfig differ
Files upstream/kernel/ksysfs.c and wsl2/kernel/ksysfs.c differ
Files upstream/lib/Kconfig and wsl2/lib/Kconfig differ
Files upstream/lib/Kconfig.debug and wsl2/lib/Kconfig.debug differ
Files upstream/lib/raid6/algos.c and wsl2/lib/raid6/algos.c differ
Only in wsl2/lib/raid6: Kconfig
Only in wsl2: Microsoft
Files upstream/net/vmw_vsock/hyperv_transport.c and wsl2/net/vmw_vsock/hyperv_transport.c differ
Only in wsl2: README-Microsoft.WSL2
```

**Comment by jadbox** [comment]
@sashalevin
`> or commits that improve hyper-v`

Can you describe briefly what these are?

**Comment by ColinIanKing** [comment]
Any chance of being a good open source citizen and getting the fixes upstreamed?

**Comment by Biswa96** [comment]
@jadbox Pass the `diff -qr` command output to `git diff` and `grep -i hyper`,you'll get the idea.

**Comment by sashalevin** [comment]
@ColinIanKing Most of the fixes are already upstream, we are planning on upstreaming all of them.

**Comment by sashalevin** [comment]
@jadbox doesn't their git commit message provide enough information about them?

**Comment by sashalevin** [comment]
@Biswa96 or just "git log v4.19.52..wsl2/master" for a list of commits.

**Comment by Biswa96** [comment]
@sashalevin Speaking of `git log`, Why this repository contains whole Linux kernel branch? This repo may contains only the patches that applied upon the 4.19 branch, like in [microsoft/azure-linux kernel](https://github.com/microsoft/azure-linux-kernel) repo. That may be easier to understand the diffenence with upstream. We, users, can just shallow clone the Linux kernel repo and `git apply` the patches.

**Comment by kdrag0n** [comment]
@Biswa96 Git commits are significantly easier to manipulate than patches. `git log` is trivial — even GitHub shows the commit history. Most other kernel repositories on GitHub include the Linux history, so why should this be an exception?

Some people may not have used patches before and just want to look at some code in this kernel. For them, it's much easier to just open the repo on GitHub and navigate to a file rather than opening the linux-stable repo, navigating to the file, and checking for changes in each of the patches in the patch repo.

**Comment by SRGOM** [comment]
@ColinianKing- Isn't Microsoft already one of the top contributors to Linux in the past few years?

**Comment by sashalevin** [comment]
@Biswa96 OTOH you can add this tree as a remote on top of the stable kernel tree and get a quick and easy way to see all the custom stuff we have on top.

Either way, moving between one way or the other is trivial, and I think it's pointless to have this "emacs vs vim" discussion here. I'll close this issue.

**Comment by billziss-gh** [comment]
@sashalevin

**Comment by sashalevin** [comment]
@billziss-gh Every commit that is not upstream has a subject and text which describes it. It's in plain English and one should be able to easily read and understand it.

**Comment by billziss-gh** [comment]
In my opinion there are many reasons that you want to document your changes outside of commit history.

1. Not everyone is familiar with git and the Linux kernel history to be able to easily identify and study your changes.
2. Even with (1) satisfied, not everyone is familiar enough with the Linux kernel internals and the discussions around it to understand your changes without some additional explanatory perspective.
3. Even with (1) and (2) satisfied, the community would save a lot of time if the effort of gathering the changes and describing them in a single place was done by the people who made them and understand them the most.
4. Even though your work was in relation to the Linux kernel, other OS kernels may wish to leverage your work and play in the WSL2 ecosystem without being Linux.

In any case, if you do not believe there is any merit in adding documentation, feel free to close the issue.

**Comment by sashalevin** [comment]
@billziss-gh I agree with your comments in general, I just don't think that there's a major change that would justify more explanation than what is provided in the commit message. We don't have a massive pile of code that "makes everything work", all our changes are minor ones and are either on their way upstream now or already are.

There is no pile of secret sauce buried under the hood that we try to keep hidden and undocumented.

I think that a separate description would make sense if we had a long series of patches that together were doing something complex that would require an explanation on it's own, but I really don't think that this is the case here.

The out of tree changes are easy to identify, github shows them as a list of commits between the latest LTS tag, and HEAD, so if you look at the project's git history (https://github.com/microsoft/WSL2-Linux-Kernel/commits/master) the commits in question are the ones starting for the commit titled "Linux 4.19.57" and the top most commit.

You will find that most of those commits are accepted upstream at this point, and for the ones that are not they are either in the process, or there is a solid explanation why they are not.

To help resolve this issue, could you point me to a commit, or a concrete logical change (that might involve a few commits) where you feel that there is not enough documentation to describe it? I'll be happy to go into details and expand the documentation in that area.

**Comment by billziss-gh** [comment]
> To help resolve this issue, could you point me to a commit, or a concrete logical change (that might involve a few commits) where you feel that there is not enough documentation to describe it?

I do not have a specific commit or commits that I need explained. I opened this issue because it was not immediately obvious what the changes were when the repository was first posted on GitHub.

My primary interest has been to understand the machinery that WSL2 uses to interface with Windows and in particular how I might be able to leverage this machinery in my own Windows driver projects.

(I am the author of a project that lets third parties write user-mode file systems for Windows. This currently has a native and a FUSE compatible API, but I have been recently working on adding full FUSE protocol support to it for native Windows processes (via `DeviceIoControl` on `\Device\WinFsp.*`) and possibly for WSL1 processes (via `read` / `write` on `/dev/fuse`). Likewise since the introduction of WSL2 I have also been exploring the possibility to have a WSL2 FUSE process expose a file system to Windows.)

> I agree with your comments in general, I just don't think that there's a major change that would justify more explanation than what is provided in the commit message.

I understand. I will now close this issue. If/when I come back to work with WSL2 specifically and I still have questions, I will post them here or in a related forum.
